clang/lib/3C/3CStats.cpp

```diff
@@ -0,0 +1,132 @@
+//=--3CStats.cpp--------------------------------------------------*- C++-*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// Implementation of all the methods in 3CStats.h
+//===----------------------------------------------------------------------===//
+
+#include "clang/3C/3CStats.h"
+#include "clang/3C/Utils.h"
+#include <time.h>
+
+void PerformanceStats::startCompileTime() {
+  CompileTimeSt = clock();
+}
+
+void PerformanceStats::endCompileTime() {
+  CompileTime += getTimeSpentInSeconds(CompileTimeSt);
+}
+
+void PerformanceStats::startConstraintBuilderTime() {
+  ConstraintBuilderTimeSt = clock();
+}
+
+void PerformanceStats::endConstraintBuilderTime() {
+  ConstraintBuilderTime += getTimeSpentInSeconds(ConstraintBuilderTimeSt);
+}
+
+void PerformanceStats::startConstraintSolverTime() {
+  ConstraintSolverTimeSt = clock();
+}
+
+void PerformanceStats::endConstraintSolverTime() {
+  ConstraintSolverTime += getTimeSpentInSeconds(ConstraintSolverTimeSt);
+}
+
+void PerformanceStats::startArrayBoundsInferenceTime() {
+  ArrayBoundsInferenceTimeSt = clock();
+}
+
+void PerformanceStats::endArrayBoundsInferenceTime() {
+  ArrayBoundsInferenceTime += getTimeSpentInSeconds(ArrayBoundsInferenceTimeSt);
+}
+
+void PerformanceStats::startRewritingTime() {
+  RewritingTimeSt = clock();
+}
+
+void PerformanceStats::endRewritingTime() {
+  RewritingTime += getTimeSpentInSeconds(RewritingTimeSt);
+}
+
+void PerformanceStats::startTotalTime() {
+  TotalTimeSt = clock();
+}
+
+void PerformanceStats::endTotalTime() {
+  TotalTime += getTimeSpentInSeconds(TotalTimeSt);
+}
+
+void PerformanceStats::incrementNumAssumeBounds() {
+  NumAssumeBoundsCasts++;
+}
+void PerformanceStats::incrementNumCheckedCasts() {
+  NumCheckedCasts++;
+}
+
+void PerformanceStats::incrementNumWildCasts() {
+  NumWildCasts++;
+}
+
+void PerformanceStats::incrementNumFixedCasts() {
+  NumFixedCasts++;
+}
+
+void PerformanceStats::incrementNumITypes() {
+  NumITypes++;
+}
+
+void PerformanceStats::incrementNumCheckedRegions() {
+  NumCheckedRegions++;
+}
+
+void PerformanceStats::incrementNumUnCheckedRegions() {
+  NumUnCheckedRegions++;
+}
```

Comment:
What about the `CHECKED_SCOPE` `#pragma` ? Anyway to notice that and compute a stat for it?

Reply:
> These stats look to only be counting stuff that 3c rewrites, not checked stuff already there unless I've missed it?

Yes. These compute those that are explicitly added by 3c.